homeassistant/components/rainforest_raven/config_flow.py

```diff
@@ -0,0 +1,159 @@
+"""Config flow for Rainforest RAVEn devices."""
+from __future__ import annotations
+
+import asyncio
+from typing import Any
+from xml.etree.ElementTree import ParseError
+
+from aioraven.serial import RAVEnSerialDevice
+import async_timeout
+from serial.serialutil import SerialException
+import serial.tools.list_ports
+from serial.tools.list_ports_common import ListPortInfo
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components import usb
+from homeassistant.const import CONF_DEVICE, CONF_MAC, CONF_NAME
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import DEFAULT_NAME, DOMAIN
+
+
+def _generate_unique_id(port: ListPortInfo) -> str:
+    """Generate unique id from usb attributes."""
+    return f"{port.vid:04X}:{port.pid:04X}_{port.serial_number}_{port.manufacturer}_{port.description}"
```

Comment:
```suggestion
def _format_id(value: str | int) -> str:
    if isinstance(value, str):
        return value

    return f"{value:04X}"


def _generate_unique_id(info: ListPortInfo | usb.UsbServiceInfo) -> str:
    """Generate unique id from usb attributes."""
    return "_".join(
        _format_id(info.vid),
        _format_id(info.pid),
        info.serial_number,
        info.manufacturer,
        info.description,
    )
```

Reply:
I did most of this in 9eba9f6, but there are a couple of changes. The VID and PID should be separated by a colon to maintain the pattern used in other extensions, and I needed to handle the `None` value on `usb.UsbServiceInfo` to avoid an exception.

In the end, I kept the f-string over the join.